File: pkg-r/tests/testthat/test-PinSource.R
local_pin_source <- function(
  data = mtcars[1:10, ],
  name = "test_data",
  type = "csv",
  ...,
  table_name = name,
  version = NULL,
  engine = NULL,
  env = parent.frame()
) {
  board <- pins::board_temp()
  suppressMessages(pins::pin_write(board, data, name, type = type, ...))
  ps <- PinSource$new(
    board,
    name,
    table_name = table_name,
    version = version,
    engine = engine
  )
  withr::defer(ps$cleanup(), envir = env)
  ps
}

describe("PinSource$new() — lazy path (parquet)", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")
  skip_if_not_installed("nanoparquet")

  it("creates proper R6 object for parquet pin", {
    ps <- local_pin_source(type = "parquet")

    expect_s3_class(ps, "PinSource")
    expect_s3_class(ps, "DBISource")
    expect_equal(ps$get_db_type(), "DuckDB")
  })

  it("executes queries against parquet pin", {
    ps <- local_pin_source(type = "parquet")

    result <- ps$execute_query("SELECT * FROM test_data")
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 10)
    expect_equal(ncol(result), ncol(mtcars))
  })

  it("filters parquet pin data correctly", {
    ps <- local_pin_source(data = mtcars, name = "cars", type = "parquet")

    result <- ps$execute_query("SELECT * FROM cars WHERE mpg > 30")
    expect_s3_class(result, "data.frame")
    expect_true(all(result$mpg > 30))
  })
})

describe("PinSource$new() — lazy path (csv)", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("reads csv pin via DuckDB file reader", {
    ps <- local_pin_source(name = "csv_data", type = "csv")

    result <- ps$execute_query("SELECT * FROM csv_data")
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 10)
  })

  it("filters csv pin data correctly", {
    ps <- local_pin_source(name = "csv_data", type = "csv")

    result <- ps$execute_query("SELECT * FROM csv_data WHERE mpg > 20")
    expect_s3_class(result, "data.frame")
    expect_true(all(result$mpg > 20))
  })
})

describe("PinSource$new() — lazy path (json)", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("reads json pin via DuckDB file reader", {
    ps <- local_pin_source(name = "json_data", type = "json")

    result <- ps$get_data()
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 10)
  })
})

describe("PinSource$new() — in-memory path (rds)", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("reads rds pin into DuckDB", {
    ps <- local_pin_source(name = "rds_data", type = "rds")

    result <- ps$execute_query("SELECT * FROM rds_data")
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 10)
  })

  it("errors when rds pin contains a non-data-frame object", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(board, list(a = 1, b = 2), "list_pin", type = "rds")
    )

    expect_error(
      PinSource$new(board, "list_pin"),
      "not a data frame"
    )
  })
})

describe("PinSource$new() — engine argument", {
  skip_if_not_installed("pins")
  skip_if_not_installed("RSQLite")

  it("uses SQLite for a data-frame (rds) pin", {
    ps <- local_pin_source(name = "rds_data", type = "rds", engine = "sqlite")

    expect_equal(ps$get_db_type(), "SQLite")
    result <- ps$execute_query("SELECT * FROM rds_data")
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 10)
  })

  it("warns and falls back to pin_read for a file-type pin", {
    expect_warning(
      ps <- local_pin_source(
        name = "csv_data",
        type = "csv",
        engine = "sqlite"
      ),
      "more efficiently"
    )

    expect_equal(ps$get_db_type(), "SQLite")
    expect_equal(nrow(ps$execute_query("SELECT * FROM csv_data")), 10)
  })
})

describe("PinSource$table_name", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("defaults to pin name", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "my_pin")

    expect_equal(ps$table_name, "my_pin")
  })

  it("can be overridden via table_name argument", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "my_pin",
      table_name = "custom_table"
    )

    expect_equal(ps$table_name, "custom_table")
    result <- ps$execute_query("SELECT * FROM custom_table")
    expect_equal(nrow(result), 5)
  })

  it("custom table_name cannot be accessed by pin name", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "my_pin",
      table_name = "custom_table"
    )

    expect_error(ps$execute_query("SELECT * FROM my_pin"))
  })
})

describe("PinSource$get_schema()", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("returns a character schema string", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "schema_test")

    schema <- ps$get_schema()
    expect_type(schema, "character")
  })

  it("schema includes table name", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "schema_test")

    expect_match(ps$get_schema(), "schema_test")
  })

  it("schema includes column names", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "schema_test")

    expect_match(ps$get_schema(), "mpg")
    expect_match(ps$get_schema(), "cyl")
  })
})

describe("PinSource$test_query()", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("returns exactly one row", {
    ps <- local_pin_source(name = "tq_test")

    result <- ps$test_query("SELECT * FROM tq_test")
    expect_s3_class(result, "data.frame")
    expect_equal(nrow(result), 1)
  })

  it("returns zero rows when query matches nothing", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "tq_test")

    result <- ps$test_query("SELECT * FROM tq_test WHERE mpg > 9999")
    expect_equal(nrow(result), 0)
  })
})

describe("PinSource$get_data_description()", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("returns a character string", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "minimal_pin")

    expect_type(ps$get_data_description(), "character")
  })

  it("includes title when set", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "described_pin",
      title = "Motor Trend Cars"
    )

    expect_match(ps$get_data_description(), "Motor Trend Cars")
  })

  it("includes description when set", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "described_pin",
      description = "Performance data for automobiles"
    )

    expect_match(ps$get_data_description(), "Performance data for automobiles")
  })

  it("includes formatted tags when set", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "tagged_pin",
      tags = c("cars", "performance")
    )

    expect_match(ps$get_data_description(), "Tags: cars, performance")
  })

  it("formats title, description, and tags together", {
    ps <- local_pin_source(
      data = mtcars[1:5, ],
      name = "full_pin",
      title = "Motor Trend Cars",
      description = "Performance data for automobiles",
      tags = c("cars", "performance")
    )

    desc <- ps$get_data_description()
    expect_match(desc, "Motor Trend Cars")
    expect_match(desc, "Performance data for automobiles")
    expect_match(desc, "Tags: cars, performance")
  })

  it("returns a string even when no explicit metadata is set", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "bare_pin")

    expect_type(ps$get_data_description(), "character")
  })
})

describe("PinSource$new() — error paths", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")
  skip_if_not_installed("nanoparquet")

  it("rejects multi-file pins", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(board, mtcars[1:5, ], "multi_pin", type = "parquet")
    )

    local_mocked_bindings(
      pin_download = function(...) c("/fake/a.parquet", "/fake/b.parquet"),
      .package = "pins"
    )
    expect_error(
      PinSource$new(board, "multi_pin"),
      "contains 2 files"
    )
  })
})

describe("PinSource DuckDB security lockdown", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("blocks external file access after data loading", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "secure_test")

    expect_error(
      ps$execute_query("SELECT * FROM read_csv_auto('/etc/passwd')")
    )
  })

  it("blocks local filesystem access", {
    ps <- local_pin_source(data = mtcars[1:5, ], name = "secure_test")

    expect_error(
      ps$execute_query("SELECT * FROM read_parquet('/tmp/any_file.parquet')")
    )
  })
})

describe("QueryChat + PinSource integration", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")

  it("auto-fills data_description from pin metadata", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(
        board,
        mtcars[1:5, ],
        "cars",
        type = "csv",
        title = "Motor Trend Cars",
        description = "Road test data"
      )
    )
    ps <- PinSource$new(board, "cars")
    qc <- QueryChat$new(data_source = ps, table_name = "cars", greeting = "Hi")
    withr::defer(qc$cleanup())

    prompt <- qc$system_prompt
    expect_match(prompt, "Motor Trend Cars")
    expect_match(prompt, "Road test data")
  })

  it("explicit data_description overrides pin metadata", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(
        board,
        mtcars[1:5, ],
        "cars",
        type = "csv",
        title = "Motor Trend Cars"
      )
    )
    ps <- PinSource$new(board, "cars")
    qc <- QueryChat$new(
      data_source = ps,
      table_name = "cars",
      greeting = "Hi",
      data_description = "Custom description"
    )
    withr::defer(qc$cleanup())

    prompt <- qc$system_prompt
    expect_match(prompt, "Custom description")
    expect_no_match(prompt, "Motor Trend Cars")
  })

  it("clears auto-filled description when switching data source", {
    skip_if_no_dataframe_engine()

    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(
        board,
        mtcars[1:5, ],
        "cars",
        type = "csv",
        title = "Motor Trend Cars"
      )
    )
    ps <- PinSource$new(board, "cars")
    qc <- QueryChat$new(data_source = ps, table_name = "cars", greeting = "Hi")

    prompt_before <- qc$system_prompt
    expect_match(prompt_before, "Motor Trend Cars")

    qc$add_table(new_test_df(), 'cars', replace = TRUE)
    prompt_after <- qc$system_prompt
    expect_no_match(prompt_after, "Motor Trend Cars")

    qc$cleanup()
  })
})

test_that("PinSource$cleanup() disconnects the connection it opened", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")
  skip_if_not_installed("nanoparquet")

  ps <- local_pin_source(type = "parquet")
  conn <- ps$conn

  ps$cleanup()

  expect_false(DBI::dbIsValid(conn))
})

describe("PinSource multi-table support", {
  skip_if_not_installed("pins")
  skip_if_not_installed("duckdb")
  skip_if_not_installed("nanoparquet")

  it("accepts a second pin via check_source_compatibility", {
    ps1 <- local_pin_source(name = "pin_a", type = "parquet")
    ps2 <- local_pin_source(name = "pin_b", type = "parquet")

    expect_no_error(
      check_source_compatibility(list(pin_a = ps1), ps2, "pin_b")
    )
  })

  it("queries across two pins through a shared DuckDB executor", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(
        board,
        data.frame(id = 1:3, x = c(10, 20, 30)),
        "pin_a",
        type = "parquet"
      )
    )
    suppressMessages(
      pins::pin_write(
        board,
        data.frame(id = 1:3, y = c("a", "b", "c")),
        "pin_b",
        type = "csv"
      )
    )

    qc <- QueryChat$new(board, "pin_a")
    withr::defer(qc$cleanup())
    qc$add_table(board, "pin_b")

    executor <- qc$.__enclos_env__$private$.table_set$executor()
    expect_s3_class(executor, "DuckDBExecutor")

    joined <- executor$execute_query(
      "SELECT a.x, b.y FROM pin_a a JOIN pin_b b ON a.id = b.id ORDER BY a.id"
    )
    expect_equal(joined$x, c(10, 20, 30))
    expect_equal(joined$y, c("a", "b", "c"))

    # Per-table schema and validation work for both pins
    schema <- executor$get_schema("pin_b", categorical_threshold = 20)
    expect_match(schema, "Table: pin_b")
    expect_no_error(
      executor$test_query(
        "SELECT * FROM pin_b",
        "pin_b",
        require_all_columns = TRUE
      )
    )
  })

  it("mixes pins and data frames in one executor", {
    ps1 <- local_pin_source(name = "pin_a", type = "parquet")
    df_source <- local_data_frame_source(new_test_df(), "test")

    executor <- build_query_executor(list(pin_a = ps1, test = df_source))
    withr::defer(executor$cleanup())

    expect_s3_class(executor, "DuckDBExecutor")
    result <- executor$execute_query(
      "SELECT COUNT(*) AS n FROM pin_a CROSS JOIN test"
    )
    expect_equal(result$n, 10 * nrow(new_test_df()))
  })

  it("materializes the snapshotted version on versioned boards", {
    board <- pins::board_temp(versioned = TRUE)
    suppressMessages(
      pins::pin_write(board, data.frame(x = 1:4), "pin_a", type = "parquet")
    )
    ps1 <- PinSource$new(board, "pin_a")
    withr::defer(ps1$cleanup())
    ps2 <- local_pin_source(name = "pin_b", type = "parquet")

    # New version of pin_a published after its PinSource was built
    suppressMessages(
      pins::pin_write(board, data.frame(x = 1), "pin_a", type = "parquet")
    )

    executor <- build_query_executor(list(pin_a = ps1, pin_b = ps2))
    withr::defer(executor$cleanup())

    expect_equal(nrow(executor$execute_query("SELECT * FROM pin_a")), 4)
  })

  it("falls back to the source's copy when the snapshot version is pruned", {
    # Non-versioned boards drop the old version on rewrite, so the version
    # snapshotted at construction no longer exists at registration time.
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(board, data.frame(x = 1:4), "pin_a", type = "parquet")
    )
    ps1 <- PinSource$new(board, "pin_a")
    withr::defer(ps1$cleanup())
    ps2 <- local_pin_source(name = "pin_b", type = "parquet")

    suppressMessages(
      pins::pin_write(board, data.frame(x = 1), "pin_a", type = "parquet")
    )

    executor <- build_query_executor(list(pin_a = ps1, pin_b = ps2))
    withr::defer(executor$cleanup())

    expect_equal(nrow(executor$execute_query("SELECT * FROM pin_a")), 4)
  })

  it("leaves pin connections open when the executor is cleaned up", {
    ps1 <- local_pin_source(name = "pin_a", type = "parquet")
    ps2 <- local_pin_source(name = "pin_b", type = "parquet")
    executor <- build_query_executor(list(pin_a = ps1, pin_b = ps2))

    # The shared connection is executor-owned; pins keep their own.
    executor$cleanup()

    expect_true(DBI::dbIsValid(ps1$conn))
    expect_true(DBI::dbIsValid(ps2$conn))
    expect_equal(nrow(ps2$execute_query("SELECT * FROM pin_b")), 10)
  })

  it("rejects adding a sqlite-engine pin to a multi-table group", {
    skip_if_not_installed("RSQLite")
    ps1 <- local_pin_source(name = "pin_a", type = "parquet")
    ps_sqlite <- suppressWarnings(
      local_pin_source(name = "pin_b", engine = "sqlite")
    )

    expect_error(
      check_source_compatibility(list(pin_a = ps1), ps_sqlite, "pin_b"),
      "engine = \"sqlite\""
    )
  })

  it("rejects adding any table when an existing pin uses sqlite", {
    skip_if_not_installed("RSQLite")
    ps_sqlite <- suppressWarnings(
      local_pin_source(name = "pin_a", engine = "sqlite")
    )
    df_source <- local_data_frame_source(new_test_df(), "test")

    expect_error(
      check_source_compatibility(list(pin_a = ps_sqlite), df_source, "test"),
      "engine = \"sqlite\""
    )
  })

  it("QueryChat$add_table() accepts a second pin", {
    board <- pins::board_temp()
    suppressMessages(
      pins::pin_write(board, mtcars[1:5, ], "pin_a", type = "parquet")
    )
    suppressMessages(
      pins::pin_write(board, mtcars[1:5, ], "pin_b", type = "parquet")
    )

    qc <- QueryChat$new(board, "pin_a")
    withr::defer(qc$cleanup())

    expect_no_error(qc$add_table(board, "pin_b"))
  })
})

File: pkg-py/tests/test_pin_source.py
from __future__ import annotations

import duckdb
import narwhals.stable.v1 as nw
import pandas as pd
import pytest

pins = pytest.importorskip("pins")

from querychat._pin_source import PinSource  # noqa: E402


@pytest.fixture
def board(tmp_path):
    return pins.board_folder(str(tmp_path))


@pytest.fixture
def sample_df():
    return pd.DataFrame(
        {
            "name": ["Alice", "Bob", "Charlie", "Diana"],
            "age": [30, 25, 35, 28],
            "score": [85.5, 92.3, 78.1, 95.0],
        }
    )


@pytest.fixture
def parquet_source(board, sample_df):
    board.pin_write(sample_df, "test_data", type="parquet")
    ps = PinSource(board, "test_data")
    yield ps
    ps.cleanup()


class TestPinSourceLazyPath:
    def test_parquet_pin(self, parquet_source):
        result = parquet_source.execute_query("SELECT * FROM test_data")
        assert isinstance(result, nw.DataFrame)
        assert len(result) == 4
        assert list(result.columns) == ["name", "age", "score"]

    def test_csv_pin(self, board, sample_df):
        board.pin_write(sample_df, "csv_data", type="csv")
        ps = PinSource(board, "csv_data")
        try:
            result = ps.execute_query("SELECT * FROM csv_data WHERE age > 28")
            assert isinstance(result, nw.DataFrame)
            assert all(result["age"] > 28)
        finally:
            ps.cleanup()

    def test_json_pin(self, board, sample_df):
        board.pin_write(sample_df, "json_data", type="json")
        ps = PinSource(board, "json_data")
        try:
            result = ps.get_data()
            assert isinstance(result, nw.DataFrame)
            assert len(result) == 4
        finally:
            ps.cleanup()

    def test_arrow_pin(self, board, sample_df):
        board.pin_write(sample_df, "arrow_data", type="arrow")
        ps = PinSource(board, "arrow_data")
        try:
            result = ps.execute_query("SELECT * FROM arrow_data")
            assert isinstance(result, nw.DataFrame)
            assert len(result) == 4
            assert list(result.columns) == ["name", "age", "score"]
        finally:
            ps.cleanup()

    def test_parquet_pin_filtered_query(self, parquet_source, sample_df):
        result = parquet_source.execute_query(
            "SELECT name FROM test_data WHERE score > 90"
        )
        assert isinstance(result, nw.DataFrame)
        assert list(result.columns) == ["name"]
        assert len(result) == 2


class TestPinSourceSchema:
    def test_get_schema(self, parquet_source):
        schema = parquet_source.get_schema(categorical_threshold=20)
        assert "test_data" in schema
        assert "name" in schema
        assert "age" in schema
        assert "score" in schema

    def test_get_schema_table_header(self, parquet_source):
        schema = parquet_source.get_schema(categorical_threshold=20)
        lines = schema.split("\n")
        assert lines[0] == "Table: test_data"
        assert lines[1] == "Columns:"

    def test_get_schema_categorical_values(self, board):
        df = pd.DataFrame(
            {"category": ["A", "B", "A", "C", "B"], "value": [1, 2, 3, 4, 5]}
        )
        board.pin_write(df, "cat_test", type="parquet")
        ps = PinSource(board, "cat_test")
        try:
            schema = ps.get_schema(categorical_threshold=5)
            assert "Categorical values:" in schema
            assert "'A'" in schema
            assert "'B'" in schema
            assert "'C'" in schema
        finally:
            ps.cleanup()

    def test_get_schema_numeric_range(self, parquet_source):
        schema = parquet_source.get_schema(categorical_threshold=20)
        assert "Range:" in schema
        assert "25" in schema
        assert "35" in schema

    def test_get_db_type(self, parquet_source):
        assert parquet_source.get_db_type() == "DuckDB"

    def test_test_query(self, parquet_source):
        result = parquet_source.test_query("SELECT * FROM test_data")
        assert len(result) == 1

    def test_test_query_trailing_semicolon(self, parquet_source):
        result = parquet_source.test_query("SELECT * FROM test_data;")
        assert len(result) == 1


class TestPinSourceMetadata:
    def test_get_data_description_with_title_and_description(self, board, sample_df):
        board.pin_write(
            sample_df,
            "described_pin",
            type="parquet",
            title="Test Dataset",
            description="A sample dataset for testing",
        )
        ps = PinSource(board, "described_pin")
        try:
            desc = ps.get_data_description()
            assert "Test Dataset" in desc
            assert "A sample dataset for testing" in desc
        finally:
            ps.cleanup()

    def test_get_data_description_minimal(self, parquet_source):
        desc = parquet_source.get_data_description()
        assert isinstance(desc, str)
        assert "test_data" in desc

    def test_pin_meta_property(self, parquet_source):
        meta = parquet_source.pin_meta
        assert meta is not None
        assert meta.name == "test_data"

    def test_pin_meta_type(self, parquet_source):
        assert parquet_source.pin_meta.type == "parquet"


class TestPinSourceTableName:
    def test_defaults_to_pin_name(self, parquet_source):
        assert parquet_source.table_name == "test_data"
        result = parquet_source.execute_query("SELECT * FROM test_data")
        assert len(result) == 4

    def test_custom_table_name(self, board, sample_df):
        board.pin_write(sample_df, "my_pin", type="parquet")
        ps = PinSource(board, "my_pin", table_name="custom_table")
        try:
            assert ps.table_name == "custom_table"
            result = ps.execute_query("SELECT * FROM custom_table")
            assert len(result) == 4
        finally:
            ps.cleanup()

    def test_custom_table_name_in_schema(self, board, sample_df):
        board.pin_write(sample_df, "schema_pin", type="parquet")
        ps = PinSource(board, "schema_pin", table_name="my_table")
        try:
            schema = ps.get_schema(categorical_threshold=10)
            assert "my_table" in schema
            assert "schema_pin" not in schema
        finally:
            ps.cleanup()


class TestPinSourceErrors:
    def test_rejects_multi_file_pin(self, board, sample_df, tmp_path):
        from unittest.mock import patch

        board.pin_write(sample_df, "multi_pin", type="parquet")

        fake_paths = [str(tmp_path / "a.parquet"), str(tmp_path / "b.parquet")]
        with (
            patch.object(board, "pin_download", return_value=fake_paths),
            pytest.raises(ValueError, match="contains 2 files"),
        ):
            PinSource(board, "multi_pin")


class TestPinSourceSecurity:
    def test_duckdb_locked_down(self, parquet_source):
        with pytest.raises(duckdb.PermissionException, match="has been disabled"):
            parquet_source.execute_query("SELECT * FROM read_csv_auto('/etc/passwd')")

    def test_blocks_unsafe_queries(self, parquet_source):
        from querychat._utils import UnsafeQueryError

        with pytest.raises(UnsafeQueryError):
            parquet_source.execute_query("DROP TABLE test_data")


class TestQueryChatPinSourceIntegration:
    def test_auto_fills_data_description(self, board, sample_df):
        from querychat import QueryChat

        board.pin_write(
            sample_df,
            "cars",
            type="parquet",
            title="Motor Trend Cars",
            description="Road test data",
        )
        ps = PinSource(board, "cars")
        qc = QueryChat(data_source=ps, table_name="cars", greeting="Hi")
        try:
            prompt = qc._table_set.system_prompt.render(qc.tools)
            assert "Motor Trend Cars" in prompt
            assert "Road test data" in prompt
        finally:
            qc.cleanup()

    def test_explicit_description_overrides_pin_metadata(self, board, sample_df):
        from querychat import QueryChat

        board.pin_write(
            sample_df,
            "cars",
            type="parquet",
            title="Motor Trend Cars",
        )
        ps = PinSource(board, "cars")
        qc = QueryChat(
            data_source=ps,
            table_name="cars",
            greeting="Hi",
            data_description="Custom description",
        )
        try:
            prompt = qc._table_set.system_prompt.render(qc.tools)
            assert "Custom description" in prompt
            assert "Motor Trend Cars" not in prompt
        finally:
            qc.cleanup()

    def test_explicit_description_survives_source_change(self, board, sample_df):
        from querychat import QueryChat

        board.pin_write(
            sample_df,
            "cars",
            type="parquet",
            title="Motor Trend Cars",
        )
        ps = PinSource(board, "cars")
        qc = QueryChat(
            data_source=ps,
            table_name="cars",
            greeting="Hi",
            data_description="Custom description",
        )
        try:
            qc.add_table(sample_df, "cars", replace=True)
            prompt = qc._table_set.system_prompt.render(qc.tools)
            assert "Custom description" in prompt
            assert "Motor Trend Cars" not in prompt
        finally:
            qc.cleanup()

    def test_clears_auto_description_on_source_change(self, board, sample_df):
        from querychat import QueryChat

        board.pin_write(
            sample_df,
            "cars",
            type="parquet",
            title="Motor Trend Cars",
        )
        ps = PinSource(board, "cars")
        qc = QueryChat(data_source=ps, table_name="cars", greeting="Hi")
        try:
            prompt_before = qc._table_set.system_prompt.render(qc.tools)
            assert "Motor Trend Cars" in prompt_before

            qc.add_table(sample_df, "cars", replace=True)
            prompt_after = qc._table_set.system_prompt.render(qc.tools)
            assert "Motor Trend Cars" not in prompt_after
        finally:
            qc.cleanup()


class TestMultiplePins:
    """Multiple pins (and pin + data frame mixes) share a DuckDB executor."""

    def test_second_pin_passes_compatibility_check(self, board, sample_df):
        from querychat._query_executor import check_source_compatibility

        board.pin_write(sample_df, "pin_a", type="parquet")
        board.pin_write(sample_df, "pin_b", type="parquet")
        first = PinSource(board, "pin_a")
        second = PinSource(board, "pin_b")
        try:
            check_source_compatibility({"pin_a": first}, second, "pin_b")
        finally:
            first.cleanup()
            second.cleanup()

    def test_two_pins_query_through_shared_executor(self, board, sample_df):
        from querychat import QueryChat
        from querychat._query_executor import DuckDBExecutor

        board.pin_write(sample_df, "pin_a", type="parquet")
        cities = pd.DataFrame(
            {"name": ["Alice", "Diana"], "city": ["Springfield", "Shelbyville"]}
        )
        board.pin_write(cities, "pin_b", type="csv")

        qc = QueryChat(board, "pin_a")
        try:
            qc.add_table(board, "pin_b")
            executor = qc._table_set.executor
            assert isinstance(executor, DuckDBExecutor)

            result = nw.from_native(
                executor.execute_query(
                    "SELECT a.name, b.city FROM pin_a a "
                    "JOIN pin_b b ON a.name = b.name ORDER BY a.name"
                )
            )
            assert result.rows(named=True) == [
                {"name": "Alice", "city": "Springfield"},
                {"name": "Diana", "city": "Shelbyville"},
            ]

            # Per-table schema and validation work for both pins
            metas = executor.get_column_metas("pin_b")
            assert {m.name for m in metas} == {"name", "city"}
            executor.test_query(
                "SELECT * FROM pin_b", table_name="pin_b", require_all_columns=True
            )
        finally:
            qc.cleanup()

    def test_pin_and_dataframe_mix(self, board, sample_df):
        from querychat import QueryChat
        from querychat._query_executor import DuckDBExecutor

        board.pin_write(sample_df, "pin_a", type="parquet")
        qc = QueryChat(board, "pin_a")
        try:
            qc.add_table(
                pd.DataFrame({"name": ["Bob"], "dept": ["Engineering"]}),
                "employees",
            )
            executor = qc._table_set.executor
            assert isinstance(executor, DuckDBExecutor)

            result = nw.from_native(
                executor.execute_query(
                    "SELECT e.dept FROM pin_a p JOIN employees e ON p.name = e.name"
                )
            )
            assert result.rows(named=True) == [{"dept": "Engineering"}]
        finally:
            qc.cleanup()

    def test_mixed_dataframe_backends_still_rejected_with_pin(self, board, sample_df):
        pl = pytest.importorskip("polars")

        from querychat import QueryChat

        board.pin_write(sample_df, "pin_a", type="parquet")
        qc = QueryChat(board, "pin_a")
        try:
            qc.add_table(sample_df, "pandas_table")
            with pytest.raises(ValueError, match="same DataFrame backend"):
                qc.add_table(pl.DataFrame({"x": [1]}), "polars_table")
        finally:
            qc.cleanup()

    def test_pin_with_non_sql_safe_name(self, board, sample_df):
        """Registry keys that need quoting work in the shared executor."""
        from querychat import QueryChat

        board.pin_write(sample_df, "sales-2026", type="parquet")
        board.pin_write(sample_df, "pin_b", type="parquet")
        qc = QueryChat(board, "sales-2026")
        try:
            qc.add_table(board, "pin_b")
            executor = qc._table_set.executor
            result = nw.from_native(
                executor.execute_query('SELECT COUNT(*) AS n FROM "sales-2026"')
            )
            assert result.rows(named=True) == [{"n": 4}]
        finally:
            qc.cleanup()

    def test_shared_executor_uses_version_resolved_at_construction(
        self, board, sample_df
    ):
        """Updating a pin after construction doesn't change the shared table."""
        from querychat import QueryChat

        board.pin_write(sample_df, "pin_a", type="parquet")
        board.pin_write(sample_df, "pin_b", type="parquet")
        qc = QueryChat(board, "pin_a")
        try:
            qc.add_table(board, "pin_b")
            # New version of pin_a published after its PinSource was built
            board.pin_write(sample_df.head(1), "pin_a", type="parquet")
            executor = qc._table_set.executor
            result = nw.from_native(
                executor.execute_query("SELECT COUNT(*) AS n FROM pin_a")
            )
            assert result.rows(named=True) == [{"n": 4}]
        finally:
            qc.cleanup()

    def test_shared_executor_falls_back_when_snapshot_version_is_pruned(
        self, tmp_path, sample_df
    ):
        """
        Non-versioned boards drop old versions on rewrite; the shared table
        must still match the source's own copy.
        """
        from querychat import QueryChat

        unversioned = pins.board_folder(str(tmp_path / "unversioned"), versioned=False)
        unversioned.pin_write(sample_df, "pin_a", type="parquet")
        unversioned.pin_write(sample_df, "pin_b", type="parquet")
        qc = QueryChat(unversioned, "pin_a")
        try:
            qc.add_table(unversioned, "pin_b")
            unversioned.pin_write(sample_df.head(1), "pin_a", type="parquet")
            executor = qc._table_set.executor
            result = nw.from_native(
                executor.execute_query("SELECT COUNT(*) AS n FROM pin_a")
            )
            assert result.rows(named=True) == [{"n": 4}]
        finally:
            qc.cleanup()

    def test_executor_cleanup_leaves_pin_connections(self, board, sample_df):
        """The shared connection is executor-owned; pins keep their own."""
        from querychat import QueryChat

        board.pin_write(sample_df, "pin_a", type="parquet")
        board.pin_write(sample_df, "pin_b", type="parquet")
        qc = QueryChat(board, "pin_a")
        try:
            qc.add_table(board, "pin_b")
            _ = qc._table_set.executor  # build the shared executor
            qc._table_set.cleanup_executor()

            # Both pins still answer source-level queries through their
            # private connections.
            for name in ("pin_a", "pin_b"):
                result = qc._data_sources[name].execute_query(
                    f"SELECT COUNT(*) AS n FROM {name}"
                )
                assert result.rows(named=True) == [{"n": 4}]
        finally:
            qc.cleanup()
